File: app/src/main/java/me/jfenn/alarmio/data/preference/TimeZonesPreferenceData.java
package me.jfenn.alarmio.data.preference;

import android.content.DialogInterface;

import java.util.Locale;
import java.util.TimeZone;

import me.jfenn.alarmio.R;
import me.jfenn.alarmio.data.PreferenceData;
import me.jfenn.alarmio.dialogs.TimeZoneChooserDialog;

public class TimeZonesPreferenceData extends CustomPreferenceData {

    private PreferenceData preference;

    public TimeZonesPreferenceData(PreferenceData preference, int title) {
        super(title);
        this.preference = preference;
    }

    @Override
    public String getValueName(ViewHolder holder) {
        int i = 0;
        for (String id : TimeZone.getAvailableIDs()) {
            if (preference.getSpecificValue(holder.getContext(), id))
                i++;
        }

        return String.format(Locale.getDefault(), holder.getContext().getString(R.string.msg_time_zones_selected), i);
    }

    @Override
    public void onClick(final ViewHolder holder) {
        TimeZoneChooserDialog dialog = new TimeZoneChooserDialog(holder.getContext());
        dialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                bindViewHolder(holder);
            }
        });
        dialog.show();
    }
}

File: app/src/main/java/me/jfenn/alarmio/data/preference/TimeZonesPreferenceData.kt
package me.jfenn.alarmio.data.preference

import me.jfenn.alarmio.R
import me.jfenn.alarmio.data.PreferenceData
import me.jfenn.alarmio.dialogs.TimeZoneChooserDialog
import java.util.*

/**
 * A preference item allowing the user to select
 * from multiple time zones (preference is a boolean,
 * should have a parameter for the zone id).
 */
class TimeZonesPreferenceData(private val preference: PreferenceData, title: Int) : CustomPreferenceData(title) {

    override fun getValueName(holder: CustomPreferenceData.ViewHolder): String {
        var i = 0
        for (id in TimeZone.getAvailableIDs()) {
            if (preference.getSpecificValue(holder.context, id))
                i++
        }

        return String.format(Locale.getDefault(), holder.context.getString(R.string.msg_time_zones_selected), i)
    }

    override fun onClick(holder: CustomPreferenceData.ViewHolder) {
        TimeZoneChooserDialog(holder.context).apply {
            setOnDismissListener { bindViewHolder(holder) }
            show()
        }
    }
}
